### ISSUE TITLE
Make prometheus operator package methods reusable by package importers

### ISSUE THREAD

**Issue by rashmichandrashekar** [prometheus-operator/prometheus-operator#6076]
**What is missing?**
Some of the methods in the prometheus operator like [validateRelabelConfig](https://github.com/prometheus-operator/prometheus-operator/blob/5f1a151a48d3ec71230c09e0ab4575b0a26560a9/pkg/prometheus/resource_selector.go#L237) are pretty useful for consumers of this package - for ex- otel-operator's [TargetAllocator](https://github.com/open-telemetry/opentelemetry-operator/tree/main/cmd/otel-allocator#target-allocator) component which uses prometheus-operator to discover and scrape CRs. Is there a way to expose this method so that the package consumers can also have the same validation as the prometheus-operator?
It will make the dependencies and upgrades a lot cleaner without having to worry about duplicating the code and copying the code when there are updates to keep the behavior consistent.

**Why do we need it?**
To make the consumers of prometheus-operator have the same behavior as the operator without having to duplicate the code and make updates easy.

**Comment by ArthurSens** [comment]
Historically the operator doesn't provide stability guarantees on its go codebase, it could happen that we make changes to the methods that downstream projects depend on without notice 😬 

That being said, I don't see a problem with us making those methods public, as long as downstream are aware of this no stability guarantees problem

**Comment by rashmichandrashekar** [comment]
> Historically the operator doesn't provide stability guarantees on its go codebase, it could happen that we make changes to the methods that downstream projects depend on without notice 😬
> 
> That being said, I don't see a problem with us making those methods public, as long as downstream are aware of this no stability guarantees problem
> 
> edit:
> 
> Just noticed the discussion in the related PR! Happy to discuss this further...Do we know of any other project that offers those guarantees? Do they have something that helps then make sure we don't introduce breaking changes?

Thanks @ArthurSens. It is not required by the prometheus-operator to be backward compatible or to guarantee stability. In this case, the otel-operator that takes depedency on the prom-operator will make sure everything works as expected when getting newer version of the prometheus-operator. @jaronoff97 - Could you pls provide your perspective here?

**Comment by simonpasquier** [comment]
Can you rephrase the description with a more generic stance and not limit it to that specific method?
We (maintainers) should agree collectively on what we're ready to support and what are the limitations for consumers. The final decision will need to be recorded in this repository for the future.

**Comment by jaronoff97** [comment]
@ArthurSens definitely understand that – our operator makes use of the servicemonitors and podmonitors so that users can swap from prometheus to otel seamlessly. We predominantly use [this method](https://github.com/prometheus-operator/prometheus-operator/blob/main/pkg/prometheus/promcfg.go#L476C34-L476C34) for doing that. We already take on the risks of no stability guarantees as I know our use of it isn't the intention of your package.

**Comment by rashmichandrashekar** [comment]
> Can you rephrase the description with a more generic stance and not limit it to that specific method? We (maintainers) should agree collectively on what we're ready to support and what are the limitations for consumers. The final decision will need to be recorded in this repository for the future.

@simonpasquier - Thanks! I have rephrased and also moved the example to the last section. Please lmk if this looks good and if not, I can update it. Also, @jaronoff97 has responded, @ArthurSens @simonpasquier - Can we please get a consensus on how we want to proceed?

**Comment by simonpasquier** [comment]
Thanks! We have bi-weekly office hours (see https://prometheus-operator.dev/docs/prologue/contributing/#office-hours-meetings) but I'm afraid that the timing (11:00 UTC) won't work for you or @jaronoff97...
Still I've added the issue to the backlog for the next meeting so we can discuss it with the other maintainers and we'll keep you posted.

**Comment by simonpasquier** [comment]
We've discussed the topic during the last office hours. The consensus among maintainers is as follows

> We're ok with code changes that allows external projects like the OpenTelemetry-Collector Operator to import `github.com/prometheus-operator/prometheus-operator`. Maintainers will not object to non-functional changes like making a function public. 
> At the same time we don’t make any stability guarantees. Though we'll refrain from introducing breaking changes without reason, our primary focus is to keep our code base healthy/easy to maintain and consuming projects should be prepared to deal with for breaking changes when they update to a new tag.
> This statement doesn't apply to the `github.com/prometheus-operator/prometheus-operator/pkg/apis/monitoring` and `github.com/prometheus-operator/prometheus-operator/pkg/client` modules.

cc @prometheus-operator/prometheus-operator-reviewers

**Comment by simonpasquier** [comment]
Regarding your initial request, it's fine to make validateRelabelConfig() a public function. But I would probably recommend that you look at the broader picture: maybe the otel collector operator could reuse higher level functions/methods to aggregate and generate the scrape configurations (for instance, [ResourceSelector](https://github.com/prometheus-operator/prometheus-operator/blob/05ea249223c81c88a7629f7599d218a4f254356d/pkg/prometheus/server/operator.go#L1486-L1509) is the key component ensuring that only valid monitoring resources are taken into account)

**Comment by jaronoff97** [comment]
Thanks for letting me know – that consensus makes sense to me and is similar to the guarantees (or lackthereof) of our own code. @rashmichandrashekar I think using a resource selector as simon described above is probably a better approach anyway.

**Comment by rashmichandrashekar** [comment]
Thanks @simonpasquier and @jaronoff97! 
@jaronoff97  - I did look into the approach of using ResourceSelector since that's the path the prometheus operator takes, but it is closely tied to namespace filtering and I dont see any implementation for that in the otel-operator repo - https://github.com/open-telemetry/opentelemetry-operator/blob/73d3cd1c1bd387bed7eac07be361defe5f2e9473/cmd/otel-allocator/watcher/promOperator.go#L56.
Also I see a couple other methods from the ResourceSelector's methods here - https://github.com/open-telemetry/opentelemetry-operator/blob/73d3cd1c1bd387bed7eac07be361defe5f2e9473/cmd/otel-allocator/watcher/promOperator.go#L290
https://github.com/open-telemetry/opentelemetry-operator/blob/73d3cd1c1bd387bed7eac07be361defe5f2e9473/cmd/otel-allocator/watcher/promOperator.go#L333
So, I want sure if this approach was tried before and the otel-operator took this approach for a reason.

**Comment by jaronoff97** [comment]
Im not sure what you mean by namespace filtering, could you elaborate?

If we need to make the target allocator act more like the prometheus operator does for this method, we should do so in order to use the ResourceSelector as stated here. I think we did what we did simply because it was what was available / thought of at the time, if there's something better from a developer perspective we should do that instead. i.e. let's give the TA more configuration / ability if it needs it in order to make the resource selector work.

**Comment by rashmichandrashekar** [comment]
Sure @jaronoff97. By namespace filtering I mean the ResourceSelector uses the [namespace selector](https://github.com/prometheus-operator/prometheus-operator/blob/05ea249223c81c88a7629f7599d218a4f254356d/pkg/prometheus/resource_selector.go#L81C4-L81C40) to pick the service monitors in a specific namespace, if it doesnt exist, it filters down to the namespace where prometheus is running which is opposite of what TA does today(i.e, look at all namespaces)

But, I agree that using ResourceSelector to get the service and pod monitors are the right way to go about this.
Looks like we need to add a few more fields to the allocator's config to propogate that to the resource selector since it assumes that it exists while calling the method. I will work on adding them.
